### DISCUSSION TITLE
Remove accidental commits from 'git subtree push'

### DISCUSSION THREAD

**Discussion by qpkorr** [community/community#21995] [New to GitHub]
<p>Hi!</p>
<p>I use git subtree for local editor (vim) plugins, including one of mine,</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/qpkorr/vim-bufkill">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/qpkorr/vim-bufkill" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/347545/181068126-38bac3f6-825a-4be8-bb43-1de26d7a6aec.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/d/2/d28e61bd6424545b65e6bddb3be83e21717e7ff1_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/d/2/d28e61bd6424545b65e6bddb3be83e21717e7ff1_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/d/2/d28e61bd6424545b65e6bddb3be83e21717e7ff1.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/d/2/d28e61bd6424545b65e6bddb3be83e21717e7ff1_2_10x10.png"></div>

<h3><a href="https://github.com/qpkorr/vim-bufkill" target="_blank" rel="noopener nofollow ugc">GitHub - qpkorr/vim-bufkill: Git repo for...</a></h3>

  <p>Git repo for http://www.vim.org/scripts/script.php?script_id=1147 - GitHub - qpkorr/vim-bufkill: Git repo for http://www.vim.org/scripts/script.php?script_id=1147</p>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>Yesterday I attempted to push some changes via<br>
git subtree push --prefix=vimfiles/bundle/bufkill vim-bufkill master<br>
but somehow it seems to have pushed not just my changes to that subtree, but also my changes to everything else not in that subtree - ie maybe 4000 commits.</p>
<p>There’s no private keys or sensitive stuff there - but I don’t think people following my repo really want to clutter their clones with my 4000 unrelated commits.</p>
<p>I was wondering about some solutions from here:</p><aside class="onebox stackexchange" data-onebox-src="https://stackoverflow.com/questions/448919/how-can-i-remove-a-commit-on-github">
  <header class="source">

      <a href="https://stackoverflow.com/questions/448919/how-can-i-remove-a-commit-on-github" target="_blank" rel="noopener nofollow ugc">stackoverflow.com</a>
  </header>

  <article class="onebox-body">
      <a href="https://stackoverflow.com/users/14755/hectorsq" target="_blank" rel="noopener nofollow ugc">
    <img alt="hectorsq" src="https://user-images.githubusercontent.com/347545/181068128-57426813-95f3-47f4-bb7e-afdca8962ba8.jpeg" class="thumbnail onebox-avatar" width="256" height="256">
  </a>

<h4>
  <a href="https://stackoverflow.com/questions/448919/how-can-i-remove-a-commit-on-github" target="_blank" rel="noopener nofollow ugc">How can I remove a commit on GitHub?</a>
</h4>

<div class="tags">
  <strong>git, github, git-commit</strong>
</div>

<div class="date">
  asked by
  
  <a href="https://stackoverflow.com/users/14755/hectorsq" target="_blank" rel="noopener nofollow ugc">
    hectorsq
  </a>
  on <a href="https://stackoverflow.com/questions/448919/how-can-i-remove-a-commit-on-github" target="_blank" rel="noopener nofollow ugc">11:20PM - 15 Jan 09 UTC</a>
</div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>
<p>
and figured since it’s not a hugely popular repo, this might be the lesser of two evils - if indeed it would work? - but I’d really appreciate some advice from those who understand this stuff far better than me.</p>
<p>Thanks in advance,<br>
John</p>

**Comment by jsoref** [comment]
<p>Generally if you <code>git push --force-with-lease $OLD_REF:$TARGET_NAME</code> for the various branches/tags to points before your push, that’ll be sufficient.</p>
<p>The accidental commits will still be reachable via github, but they won’t generally appear for people using <code>git fetch</code> or <code>git clone</code>.</p>

**Comment by qpkorr** [comment]
<p>After doing more research, I found this SO answer which begins:</p>
<blockquote>
<p>My conclusion, after many many hours of struggling (reading, discussing, retrying in different ways, etc) was that git-subtree just doesn’t work properly.</p>
</blockquote>
<aside class="onebox stackexchange" data-onebox-src="https://stackoverflow.com/questions/61150709/git-subtree-push-seems-to-push-commits-that-dont-apply-to-the-subtree/62615285#62615285">
  <header class="source">

      <a href="https://stackoverflow.com/questions/61150709/git-subtree-push-seems-to-push-commits-that-dont-apply-to-the-subtree/62615285#62615285" target="_blank" rel="noopener nofollow ugc">stackoverflow.com</a>
  </header>

  <article class="onebox-body">
      <a href="https://stackoverflow.com/users/136785/metal450" target="_blank" rel="noopener nofollow ugc">
    <img alt="Metal450" src="https://user-images.githubusercontent.com/347545/181068129-8cb77d47-8f12-4d68-bf51-0a653cad719d.jpeg" class="thumbnail onebox-avatar" width="256" height="256">
  </a>

<h4>
  <a href="https://stackoverflow.com/questions/61150709/git-subtree-push-seems-to-push-commits-that-dont-apply-to-the-subtree/62615285#62615285" target="_blank" rel="noopener nofollow ugc">Git subtree push seems to push commits that don't apply to the subtree?</a>
</h4>

<div class="tags">
  <strong>git, git-subtree</strong>
</div>

<div class="date">
  
  answered by
  <a href="https://stackoverflow.com/users/136785/metal450" target="_blank" rel="noopener nofollow ugc">
    Metal450
  </a>
  on <a href="https://stackoverflow.com/questions/61150709/git-subtree-push-seems-to-push-commits-that-dont-apply-to-the-subtree/62615285#62615285" target="_blank" rel="noopener nofollow ugc">08:35PM - 27 Jun 20 UTC</a>
</div>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>It seems git subtree doesn’t work the way I - and the author of the Atlassian post about it think it should (or used to?):</p><aside class="onebox allowlistedgeneric" data-onebox-src="https://www.atlassian.com/git/tutorials/git-subtree">
  <header class="source">
      <img src="https://user-images.githubusercontent.com/347545/181068130-b86e8d86-07f6-4298-8e95-1f73fe6720c1.png" class="site-icon" width="32" height="32">

      <a href="https://www.atlassian.com/git/tutorials/git-subtree" target="_blank" rel="noopener nofollow ugc">Atlassian</a>
  </header>

  <article class="onebox-body">
    

<h3><a href="https://www.atlassian.com/git/tutorials/git-subtree" target="_blank" rel="noopener nofollow ugc">Git subtree: the alternative to Git submodule | Atlassian Git Tutorial</a></h3>

  <p>Git subtree lets you nest one repository inside another as a sub-directory. It is one of several ways Git projects can manage project dependencies. Learn h</p>


  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<p>I figured I might as well try the SO suggestion of rolling back the head of my repo and force pushing - though git subtree doesn’t support force pushing but changing line 1017 of /usr/lib/git-core/git-subtree from<br>
git push “$repository” “$localrev”:“refs/heads/$remoteref”<br>
to<br>
git push “$repository” “$localrev”:“refs/heads/$remoteref” --force<br>
did allow me to remove one commit from github history. Unfortunately, that still left 4082 commits, not the 4083 I used to have - ie removing the merge commit didn’t remove all the extra junk commits.</p>
<p>A new repo is looking increasingly necessary I think.</p>

**Comment by qpkorr** [comment, marked as answer]
<p>I signed up for a paid github account to submit a support request about this, and their solution, which I’ve implemented, was</p>
<ul>
<li>use the link builder below, filling in the appropriate bits, to view your repository at the last commit before everything went wrong:<br>
<a href="https://github.com/USER/REPO/tree/SHA" rel="noopener nofollow ugc">https://github.com/USER/REPO/tree/SHA</a><br>
In my case - there were about 4000 erroneous commits sprinkled over years, so I struggled with this step. Github support were able to access their push logs to narrow down where to look, and provided the best link for me</li>
<li>from that link, click on the branch selector in the top left corner with a branch icon and the SHA and a drop down arrow</li>
<li>when the drop down appears, write a new branch name, eg <code>restored</code>, then click “Create branch: <code>restored</code> from ‘<code>SHA</code>’”.</li>
<li>in my case I had one extra commit from <code>master</code> that I wanted, so I cloned the repo locally again (<code>git clone ...</code>), got all branches (<code>git pull</code>), moved to the <code>restored</code> remote branch (<code>git checkout --track origin/restored</code>), then cherry-picked from the bloated <code>master</code> branch (<code>git cherry-pick master</code>). I was a bit surprised that this grabbed only the relevant commit from all the mess that was my master branch. Then push back to the github ‘restored’ branch (<code>git push</code>).</li>
<li>then make the new <code>restored</code> branch the <a href="https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/changing-the-default-branch" rel="noopener nofollow ugc">the default branch of the repository</a>.</li>
<li>then <a href="https://www.git-tower.com/learn/git/faq/github-delete-branch" rel="noopener nofollow ugc">delete the master branch</a>
</li>
<li>and finally <a href="https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-branches-in-your-repository/renaming-a-branch" rel="noopener nofollow ugc">rename <code>restored</code> back to <code>master</code></a>
</li>
</ul>
<p>I figure I best add a warning in my README.md file pointing people here in case I’ve caused problems for people with forks - but I’m hopeful this is the best solution available.</p>
